runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WindmillStateInternals.java

```diff
@@ -1588,7 +1599,522 @@ private Future<Iterable<Map.Entry<ByteString, V>>> getFuture() {
+      localAdditions = ArrayListMultimap.create();
+      localRemovals = Sets.newHashSet();
+      cleared = true;
+      complete = true;
+      allKeysKnown = true;
+    }
+
+    @Override
+    public ReadableState<Iterable<K>> keys() {
+      return new ReadableState<Iterable<K>>() {
+        @Override
+        public Iterable<K> read() {
+          if (allKeysKnown) {
+            return Iterables.unmodifiableIterable(
+                Iterables.transform(
+                    Iterables.filter(
+                        keyStateMap.values(),
+                        keyState -> keyState.existence == KeyExistence.KNOWN_EXIST),
+                    keyState -> keyState.originalKey));
+          }
+          Future<Iterable<Entry<ByteString, Iterable<V>>>> persistedData = getFuture(true);
+          try (Closeable scope = scopedReadState()) {
+            Iterable<Entry<ByteString, Iterable<V>>> entries = persistedData.get();
+            Iterable<Entry<Object, K>> keys =
+                Iterables.transform(
+                    entries,
+                    entry -> {
+                      try {
+                        K originalKey = keyCoder.decode(entry.getKey().newInput());
+                        return new AbstractMap.SimpleEntry<>(
+                            keyCoder.structuralValue(originalKey), originalKey);
+                      } catch (IOException e) {
+                        throw new RuntimeException(e);
+                      }
+                    });
+            keys =
+                Iterables.filter(
+                    keys,
+                    entry -> {
+                      KeyState keyState = keyStateMap.getOrDefault(entry.getKey(), null);
+                      // this is a key that exists in windmill but is not cached.
+                      if (keyState == null) return true;
+                      // if the key is marked as deleted in cache, ignore it.
+                      return keyState.existence != KeyExistence.KNOWN_NONEXISTENT;
+                    });
+            if (entries instanceof Weighted) {
+              // This is a known amount of data, cache them all.
+              keys.forEach(
+                  entry -> {
+                    KeyState keyState =
+                        keyStateMap.computeIfAbsent(
+                            entry.getKey(), stk -> new KeyState(entry.getValue()));
+                    keyState.existence = KeyExistence.KNOWN_EXIST;
+                  });
+              allKeysKnown = true;
+              keyStateMap
+                  .values()
+                  .removeIf(keyState -> keyState.existence != KeyExistence.KNOWN_EXIST);
+              return Iterables.unmodifiableIterable(
+                  Iterables.transform(keyStateMap.values(), keyState -> keyState.originalKey));
+            } else {
+              return Iterables.unmodifiableIterable(
+                  Iterables.concat(
+                      // This is the part of keys that are cached.
```

Comment:
just checking that we don't need to provide ordering guarantees.  Otherwise we should interleave the cache and fetched results.